packages/cli-utils/src/index.ts

```diff
@@ -12,11 +12,145 @@ import type { GraphQLSPConfig } from './lsp';
 import { ensureTadaIntrospection } from './tada';
 
+interface GenerateSchemaOptions {
+  headers?: Record<string, string>;
+  output?: string;
+  cwd?: string;
+}
+
+export async function generateSchema(
+  target: string,
+  { headers, output, cwd = process.cwd() }: GenerateSchemaOptions
+) {
+  let url: URL | undefined;
+
+  try {
+    url = new URL(target);
+  } catch (e) {}
+
+  let introspection: IntrospectionQuery;
+  if (url) {
+    const response = await fetch(url!.toString(), {
+      method: 'POST',
+      headers: {
+        ...headers,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        query: getIntrospectionQuery({
+          descriptions: true,
+          schemaDescription: false,
+          inputValueDeprecation: false,
+          directiveIsRepeatable: false,
+          specifiedByUrl: false,
+        }),
+      }),
+    });
+
+    if (response.ok) {
+      const text = await response.text();
+
+      try {
+        const result = JSON.parse(text);
+        if (result.data) {
+          introspection = (result as { data: IntrospectionQuery }).data;
+        } else {
+          console.error(`Got invalid response ${JSON.stringify(result)}`);
+          return;
+        }
+      } catch (e) {
+        console.error(`Got invalid JSON ${text}`);
+        return;
+      }
+    } else {
+      console.error(`Got invalid response ${await response.text()}`);
+      return;
+    }
+  } else {
+    const path = resolve(cwd, target);
+    const fileContents = await fs.readFile(path, 'utf-8');
+
+    try {
+      introspection = JSON.parse(fileContents);
+    } catch (e) {
+      console.error(`Got invalid JSON ${fileContents}`);
+      return;
+    }
+  }
+
+  const schema = buildClientSchema(introspection!);
+
+  let destination = output;
+  if (!destination) {
+    const tsconfigpath = path.resolve(cwd, 'tsconfig.json');
+    const hasTsConfig = existsSync(tsconfigpath);
+    if (!hasTsConfig) {
+      console.error(`Could not find a tsconfig in the working-directory.`);
+      return;
+    }
+
+    const tsconfigContents = await fs.readFile(tsconfigpath, 'utf-8');
+    let tsConfig: TsConfigJson;
+    try {
+      tsConfig = parse(tsconfigContents) as TsConfigJson;
+    } catch (err) {
+      console.error(err);
+      return;
+    }
+
+    if (!hasGraphQLSP(tsConfig)) {
+      console.error(`Could not find a "@0no-co/graphqlsp" plugin in your tsconfig.`);
+      return;
+    }
+
+    const foundPlugin = tsConfig.compilerOptions!.plugins!.find(
+      (plugin) => plugin.name === '@0no-co/graphqlsp'
+    ) as GraphQLSPConfig;
+
+    destination = foundPlugin.schema;
+
+    if (!foundPlugin.schema.endsWith('.graphql')) {
+      console.error(`Found "${foundPlugin.schema}" which is not a path to a GraphQL Schema.`);
+      return;
+    }
+  }
+
+  await fs.writeFile(resolve(cwd, destination), printSchema(schema), 'utf-8');
+}
+
+export async function generateOutput(cwd: string = process.cwd()) {
```

Comment:
Let's rename this to `generateTadaTypes` or something that matches the CLI command more